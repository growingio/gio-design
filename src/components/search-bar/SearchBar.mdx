import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import SearchBar from './index';

# Search Bar 搜索框

<Subtitle>在所有内容中，通过输入内容的关键信息筛选出某个内容。</Subtitle>

## 参数说明

<ArgsTable of={SearchBar} />

## 代码演示

### 默认

[Example](/?path=/story/functional-components-searchbar--default)
