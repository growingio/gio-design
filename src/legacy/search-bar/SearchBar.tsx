import React, { useState } from 'react';
import classnames from 'classnames';
import { CloseCircleFilled, SearchOutlined } from '@gio-design/icons';
import { usePrefixCls, useLocale } from '@gio-design/utils';
import { Button } from '../../index';
import Input from '../input';
import { SearchBarProps } from './interfaces';
import defaultLocale from './locales/zh-CN';

export { SearchBarProps } from './interfaces';

const getStorage = (key: string): string[] => {
  const empty: string[] = [];
  try {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return localStorage.getItem(key) ? JSON.parse(localStorage.getItem(key)) : empty;
  } catch (error) {
    return empty;
  }
};

const setStorage = (key: string, value: string) => {
  try {
    const oldValue = getStorage(key);
    if (!value || oldValue.some((item) => item === value)) {
      return oldValue;
    }
    const newValue = oldValue.concat(value);
    localStorage.setItem(key, JSON.stringify(newValue));
    return newValue;
  } catch (error) {
    return [];
  }
};
const findStorage = (key: string, value: string): string[] => {
  const storages = getStorage(key);
  return storages.filter((item) => item.startsWith(value));
};

const clearStorage = (key: string): string[] => {
  localStorage.removeItem(key);
  return [];
};

const SearchBar: React.FC<SearchBarProps> = (props: SearchBarProps) => {
  const prefixCls = usePrefixCls('searchbar-legacy');
  const [searchValue, setSearchValue] = useState('');
  const {
    showStorage = false,
    storageNum = 5,
    allowClearStorage = false,
    showClear = false,
    disabled = false,
    size = 'middle',
    inputStyle,
    style,
    placeholder,
    value,
    onChange = setSearchValue,
    id,
    className,
  } = props;
  const storageKey = React.useMemo(() => `${prefixCls}-storage-${id}`, [id, prefixCls]);

  const [searchStorage, setSearchStorage] = React.useState(getStorage(storageKey));
  const [showDropdown, setShowDropdown] = React.useState(false);

  const locale = useLocale('SearchBar');
  const { show, record, clearText }: { show: string; record: string; clearText: string } = {
    ...defaultLocale,
    ...locale,
  };

  const handleClearStorage = () => {
    const emptyValue = clearStorage(storageKey);
    setSearchStorage(emptyValue);
  };

  const handleClearValue = () => {
    onChange('');
    const newValue = findStorage(storageKey, '');
    setSearchStorage(newValue);
  };

  const handleFocus = () => {
    searchStorage.length && setShowDropdown(true);
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onChange(e.target.value);
    setShowDropdown(false);
  };

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    // eslint-disable-next-line @typescript-eslint/no-shadow
    const { value } = e.target;
    setTimeout(() => {
      id && setStorage(storageKey, value);
      setShowDropdown(false);
    }, 200);
  };

  const renderSuffix = () => {
    if (value || searchValue) {
      return showClear ? (
        <CloseCircleFilled className={`${prefixCls}-suffix-close`} onClick={handleClearValue} />
      ) : null;
    }
    return <SearchOutlined className={`${prefixCls}-suffix-search`} />;
  };

  // 按esc建关闭下拉框
  const handleKeyUp = (e: any) => {
    e.keyCode === 27 && handleBlur(e);
  };

  const renderStorage = () => {
    if (!showStorage || !showDropdown) {
      return null;
    }

    return (
      <div className={`${prefixCls}-dropdown`}>
        {allowClearStorage && (
          <div className={`${prefixCls}-dropdown-clear`}>
            <span className={`${prefixCls}-dropdown-clear-text`}>
              {show}
              {storageNum}
              {record}
            </span>
            <Button type="text" onClick={handleClearStorage}>
              {clearText}
            </Button>
          </div>
        )}
        {searchStorage
          .slice(searchStorage.length - storageNum >= 0 ? searchStorage.length - storageNum : 0, searchStorage.length)
          .reverse()
          .map((item) => (
            <div
              onClick={() => {
                onChange(item);
              }}
              className={`${prefixCls}-dropdown-item`}
              key={item}
              aria-hidden="true"
            >
              {item}
            </div>
          ))}
      </div>
    );
  };

  const wrapperCls = classnames(prefixCls, className);
  return (
    <div className={wrapperCls} style={style}>
      <Input
        disabled={disabled}
        size={size}
        style={inputStyle}
        suffix={renderSuffix()}
        value={value || searchValue}
        placeholder={placeholder}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onKeyUp={handleKeyUp}
      />
      {renderStorage()}
    </div>
  );
};

export default SearchBar;
