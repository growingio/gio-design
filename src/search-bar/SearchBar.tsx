import { CloseCircleFilled, SearchOutlined } from '@gio-design/icons';
import { usePrefixCls } from '@gio-design/utils';
import classNames from 'classnames';
import React, { useCallback, useMemo, useState } from 'react';
import Input from '../input/Input';
import { SearchBarProps } from './interface';

const SearchBar = React.forwardRef<HTMLInputElement, SearchBarProps>((props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    onChange: onChangeFC,
    value: enterValue,
    disabled,
    placeholder,
    onSearch,
  } = props;

  const prefixCls = usePrefixCls('search', customizePrefixCls);
  const [value, setValue] = useState(enterValue);

  const [canClear, setClear] = useState(!!enterValue);

  const onClear = useCallback(() => {
    if (!disabled) {
      onSearch?.('');
      setClear(false);
      setValue('');
    }
  }, [onSearch, disabled]);

  const suffixCls = useMemo(
    () =>
      classNames(`${prefixCls}__suffix`, {
        [`${prefixCls}__suffix-clear`]: canClear,
        [`${prefixCls}__suffix-disabled`]: disabled,
      }),
    [prefixCls, canClear, disabled]
  );

  const suffix = useMemo(
    () =>
      canClear && !disabled ? (
        <CloseCircleFilled className={suffixCls} onClick={onClear} />
      ) : (
        <SearchOutlined className={suffixCls} />
      ),
    [suffixCls, canClear, onClear, disabled]
  );

  const onChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const inputValue = e.target.value;
      onChangeFC?.(e);
      onSearch?.(inputValue);
      setValue(inputValue);
      setClear(!!inputValue);
    },
    [onChangeFC, onSearch]
  );

  return (
    <Input
      data-testid="search-bar"
      {...props}
      type="text"
      placeholder={placeholder}
      value={value}
      onChange={onChange}
      suffix={suffix}
      ref={ref}
    />
  );
});

export default SearchBar;
