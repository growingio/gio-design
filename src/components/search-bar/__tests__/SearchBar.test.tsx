import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Default } from '../SearchBar.stories';
import SearchBar from '../index';

describe('Testing search-bar', () => {
  it('basic search-bar', () => {
    const { container } = render(<Default {...Default.args} />);
    fireEvent.change(screen.getByRole('textbox'), { target: { value: '随便输点东西' } });
    expect(container.getElementsByClassName('gio-input__content')[0].getAttribute('value')).toEqual('随便输点东西');

    fireEvent.keyDown(container.getElementsByClassName('gio-input__content')[0], { key: 'Enter' });
    fireEvent.click(screen.getByRole('img'));
    expect(container.getElementsByClassName('gio-input__content')[0].getAttribute('value')).toEqual('');
  });

  it('should storage the search value', () => {
    const { container } = render(
      <div>
        <SearchBar
          size="middle"
          placeholder="请搜索..."
          storageNum={5}
          disabled={false}
          showStorage
          showClear
          allowClearStorage
        />
      </div>
    );
    fireEvent.focus(screen.getByRole('textbox'));
    fireEvent.change(screen.getByRole('textbox'), { target: { value: '123456' } });

    fireEvent.click(screen.getByRole('img'));
    fireEvent.blur(screen.getByRole('textbox'));
    fireEvent.focus(screen.getByRole('textbox'));
    screen.logTestingPlaygroundURL();
  });
});
