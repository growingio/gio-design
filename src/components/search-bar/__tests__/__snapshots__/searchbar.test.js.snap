// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should support size 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-searchbar",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "gio-input gio-input--large gio-input__container",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "gio-input__content gio-input__content-suffix",
              "placeholder": "",
              "type": "text",
              "value": "",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "gio-input__suffix",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "aria-label": "search-outlined",
                    "class": "gio-icon gio-searchbar-suffix-search",
                    "role": "img",
                  },
                  "children": Array [
                    Object {
                      "attribs": Object {
                        "fill": "currentColor",
                        "height": "16px",
                        "viewBox": "0 0 14.01 14",
                        "width": "16px",
                        "xmlns": "http://www.w3.org/2000/svg",
                      },
                      "children": Array [
                        Object {
                          "attribs": Object {},
                          "children": Array [
                            Object {
                              "attribs": Object {},
                              "children": Array [],
                              "name": "style",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "style",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                          ],
                          "name": "defs",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": Object {
                            "attribs": Object {},
                            "children": Array [
                              Object {
                                "attribs": Object {},
                                "children": Array [
                                  Object {
                                    "attribs": Object {
                                      "d": "M6 1a5 5 0 11-5 5 5 5 0 015-5zm4.48 10.19a1 1 0 00-1.28-.13A5.83 5.83 0 016 12a6 6 0 115.07-2.79 1 1 0 00.13 1.26c.93 1 2.65 2.68 2.65 2.68a.49.49 0 01-.7.7z",
                                      "fill": "currentColor",
                                      "fill-rule": "evenodd",
                                      "id": "search-outlined_svg__search",
                                    },
                                    "children": Array [],
                                    "name": "path",
                                    "namespace": "http://www.w3.org/2000/svg",
                                    "next": null,
                                    "parent": [Circular],
                                    "prev": null,
                                    "type": "tag",
                                    "x-attribsNamespace": Object {
                                      "d": undefined,
                                      "fill": undefined,
                                      "fill-rule": undefined,
                                      "id": undefined,
                                    },
                                    "x-attribsPrefix": Object {
                                      "d": undefined,
                                      "fill": undefined,
                                      "fill-rule": undefined,
                                      "id": undefined,
                                    },
                                  },
                                ],
                                "name": "g",
                                "namespace": "http://www.w3.org/2000/svg",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {},
                                "x-attribsPrefix": Object {},
                              },
                            ],
                            "name": "g",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": null,
                            "parent": [Circular],
                            "prev": [Circular],
                            "type": "tag",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {},
                          "x-attribsPrefix": Object {},
                        },
                        Object {
                          "attribs": Object {},
                          "children": Array [
                            Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "d": "M6 1a5 5 0 11-5 5 5 5 0 015-5zm4.48 10.19a1 1 0 00-1.28-.13A5.83 5.83 0 016 12a6 6 0 115.07-2.79 1 1 0 00.13 1.26c.93 1 2.65 2.68 2.65 2.68a.49.49 0 01-.7.7z",
                                    "fill": "currentColor",
                                    "fill-rule": "evenodd",
                                    "id": "search-outlined_svg__search",
                                  },
                                  "children": Array [],
                                  "name": "path",
                                  "namespace": "http://www.w3.org/2000/svg",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "d": undefined,
                                    "fill": undefined,
                                    "fill-rule": undefined,
                                    "id": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "d": undefined,
                                    "fill": undefined,
                                    "fill-rule": undefined,
                                    "id": undefined,
                                  },
                                },
                              ],
                              "name": "g",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                          ],
                          "name": "g",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": null,
                          "parent": [Circular],
                          "prev": Object {
                            "attribs": Object {},
                            "children": Array [
                              Object {
                                "attribs": Object {},
                                "children": Array [],
                                "name": "style",
                                "namespace": "http://www.w3.org/2000/svg",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "style",
                                "x-attribsNamespace": Object {},
                                "x-attribsPrefix": Object {},
                              },
                            ],
                            "name": "defs",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": [Circular],
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                          "type": "tag",
                          "x-attribsNamespace": Object {},
                          "x-attribsPrefix": Object {},
                        },
                      ],
                      "name": "svg",
                      "namespace": "http://www.w3.org/2000/svg",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "fill": undefined,
                        "height": undefined,
                        "viewBox": undefined,
                        "width": undefined,
                        "xmlns": "http://www.w3.org/2000/xmlns/",
                      },
                      "x-attribsPrefix": Object {
                        "fill": undefined,
                        "height": undefined,
                        "viewBox": undefined,
                        "width": undefined,
                        "xmlns": "",
                      },
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "aria-label": undefined,
                    "class": undefined,
                    "role": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "aria-label": undefined,
                    "class": undefined,
                    "role": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "gio-input__suffix",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "aria-label": "search-outlined",
                  "class": "gio-icon gio-searchbar-suffix-search",
                  "role": "img",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "fill": "currentColor",
                      "height": "16px",
                      "viewBox": "0 0 14.01 14",
                      "width": "16px",
                      "xmlns": "http://www.w3.org/2000/svg",
                    },
                    "children": Array [
                      Object {
                        "attribs": Object {},
                        "children": Array [
                          Object {
                            "attribs": Object {},
                            "children": Array [],
                            "name": "style",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "style",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                        ],
                        "name": "defs",
                        "namespace": "http://www.w3.org/2000/svg",
                        "next": Object {
                          "attribs": Object {},
                          "children": Array [
                            Object {
                              "attribs": Object {},
                              "children": Array [
                                Object {
                                  "attribs": Object {
                                    "d": "M6 1a5 5 0 11-5 5 5 5 0 015-5zm4.48 10.19a1 1 0 00-1.28-.13A5.83 5.83 0 016 12a6 6 0 115.07-2.79 1 1 0 00.13 1.26c.93 1 2.65 2.68 2.65 2.68a.49.49 0 01-.7.7z",
                                    "fill": "currentColor",
                                    "fill-rule": "evenodd",
                                    "id": "search-outlined_svg__search",
                                  },
                                  "children": Array [],
                                  "name": "path",
                                  "namespace": "http://www.w3.org/2000/svg",
                                  "next": null,
                                  "parent": [Circular],
                                  "prev": null,
                                  "type": "tag",
                                  "x-attribsNamespace": Object {
                                    "d": undefined,
                                    "fill": undefined,
                                    "fill-rule": undefined,
                                    "id": undefined,
                                  },
                                  "x-attribsPrefix": Object {
                                    "d": undefined,
                                    "fill": undefined,
                                    "fill-rule": undefined,
                                    "id": undefined,
                                  },
                                },
                              ],
                              "name": "g",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "tag",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                          ],
                          "name": "g",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": null,
                          "parent": [Circular],
                          "prev": [Circular],
                          "type": "tag",
                          "x-attribsNamespace": Object {},
                          "x-attribsPrefix": Object {},
                        },
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                      Object {
                        "attribs": Object {},
                        "children": Array [
                          Object {
                            "attribs": Object {},
                            "children": Array [
                              Object {
                                "attribs": Object {
                                  "d": "M6 1a5 5 0 11-5 5 5 5 0 015-5zm4.48 10.19a1 1 0 00-1.28-.13A5.83 5.83 0 016 12a6 6 0 115.07-2.79 1 1 0 00.13 1.26c.93 1 2.65 2.68 2.65 2.68a.49.49 0 01-.7.7z",
                                  "fill": "currentColor",
                                  "fill-rule": "evenodd",
                                  "id": "search-outlined_svg__search",
                                },
                                "children": Array [],
                                "name": "path",
                                "namespace": "http://www.w3.org/2000/svg",
                                "next": null,
                                "parent": [Circular],
                                "prev": null,
                                "type": "tag",
                                "x-attribsNamespace": Object {
                                  "d": undefined,
                                  "fill": undefined,
                                  "fill-rule": undefined,
                                  "id": undefined,
                                },
                                "x-attribsPrefix": Object {
                                  "d": undefined,
                                  "fill": undefined,
                                  "fill-rule": undefined,
                                  "id": undefined,
                                },
                              },
                            ],
                            "name": "g",
                            "namespace": "http://www.w3.org/2000/svg",
                            "next": null,
                            "parent": [Circular],
                            "prev": null,
                            "type": "tag",
                            "x-attribsNamespace": Object {},
                            "x-attribsPrefix": Object {},
                          },
                        ],
                        "name": "g",
                        "namespace": "http://www.w3.org/2000/svg",
                        "next": null,
                        "parent": [Circular],
                        "prev": Object {
                          "attribs": Object {},
                          "children": Array [
                            Object {
                              "attribs": Object {},
                              "children": Array [],
                              "name": "style",
                              "namespace": "http://www.w3.org/2000/svg",
                              "next": null,
                              "parent": [Circular],
                              "prev": null,
                              "type": "style",
                              "x-attribsNamespace": Object {},
                              "x-attribsPrefix": Object {},
                            },
                          ],
                          "name": "defs",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": [Circular],
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {},
                          "x-attribsPrefix": Object {},
                        },
                        "type": "tag",
                        "x-attribsNamespace": Object {},
                        "x-attribsPrefix": Object {},
                      },
                    ],
                    "name": "svg",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "fill": undefined,
                      "height": undefined,
                      "viewBox": undefined,
                      "width": undefined,
                      "xmlns": "http://www.w3.org/2000/xmlns/",
                    },
                    "x-attribsPrefix": Object {
                      "fill": undefined,
                      "height": undefined,
                      "viewBox": undefined,
                      "width": undefined,
                      "xmlns": "",
                    },
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "aria-label": undefined,
                  "class": undefined,
                  "role": undefined,
                },
                "x-attribsPrefix": Object {
                  "aria-label": undefined,
                  "class": undefined,
                  "role": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "gio-input__content gio-input__content-suffix",
                "placeholder": "",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
